### PR TITLE
Feature/Contacts and StudentContactAssociations

### PR DESCRIPTION
**Pull request by sleblanc23**
## Description & motivation
In Data Standard v5.0, Parents were renamed to Contacts. The new Contact entity is identical to the previous Parent entity with the addition of an optional GenderIdentity, PreferredFirstName, and PreferredLastName fields. StudentParentAssociations was also renamed to StudentContactAssociations and the parentReference in SurveyResponses was renamed to contactReference.

This PR adds base and stage models for the new resources, which union in the data from the deprecated parent-named resources.

## Breaking changes introduced by this PR:
- `stg_ef3__survey_responses` now includes `k_contact` instead of `k_parent`.
- Three new columns were added to the base and stage models for parents. These will always be null for parents but may be populated in contacts, so they are required for the union operation.

## PR Merge Priority:
- Medium

## Changes to existing files:
- `gen_skey` : add `k_contact` to key definitions
- `base_ef3__parents` : add three new fields which were introduced along with the Contacts rename in order to allow the tables to be unioned in stage
- `base_ef3__survey_responses` : coalesce parentReference and contactReference into a single `contact_reference`
- `stg_ef3__survey_responses` : rename k_parent to k_contact

## New files created:
Base and stage models for `contacts`, `student_contact_associations`, and unpacked `contacts` lists.

## Tests and QC done:
Ran in Stadium SCDE; confirmed that columns were populating correctly (where data is available) and staging tables included both Parent and Contact data. 

## Future ToDos & Questions:
- I think the impact of the switching to `k_contact` in `stg_ef3__survey_responses` is small, possibly none - we don't have any warehouse models for surveys yet and I didn't see the field referenced in any implementation repos.

### CODE REVIEW

File: models/staging/edfi_3/stage/stg_ef3__contacts.sql [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                'lower(contact_unique_id)'
+            ]
+        ) }} as k_contact,
+        unioned.*
+        {{ extract_extension(model_name=this.name, flatten=True) }}
+    from unioned
+),
+deduped as (
+    {{
+        dbt_utils.deduplicate(
```
Comment (rlittle08) [inline models/staging/edfi_3/stage/stg_ef3__contacts.sql:30]: this looks correct but might be nice to test this out in SC to see edge case behavior where there's overlapping parents & contacts (if there are any)

File: models/staging/edfi_3/base/base_ef3__parents.sql [hunk 2]
```diff
@@ -20,6 +20,11 @@ renamed as (
         v:maidenName::string                                                             as maiden_name,
         v:generationCodeSuffix::string                                                   as generation_code_suffix,
         v:personalTitlePrefix::string                                                    as personal_title_prefix,
+        -- the following three fields were introduced to the Contacts resource, which replaced Parents in v5.0
```
[the remark this answers is not in the document]
Comment (sleblanc23) [reply models/staging/edfi_3/base/base_ef3__parents.sql:23]: We could, but I think it would be more awkward because union_relations requires the entities you're unioning to be tables or materialized views, not CTEs. I would need to do the rename from parent_unique_id to contact_unique_id either here (probably too breaking) or in a separate model somewhere before using union_relations. This seemed cleaner to me.

File: models/staging/edfi_3/stage/stg_ef3__contacts.sql (hunk 1 above)
Comment (sleblanc23) [reply models/staging/edfi_3/stage/stg_ef3__contacts.sql:30]: SC didn't have any overlapping parents and contacts because they made a change to remove a leading zero from their IDs. I think ideally they would be the same to avoid duplicates in dim_parent

File: models/staging/edfi_3/stage/stg_ef3__student_contact_associations.sql [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        -- we can't use the gen_skey macro here because we're bringing in the deprecated parents endpoint data, which contains a parentReference that won't work
+        iff(
+            contact_unique_id is not null, 
+            md5(cast(coalesce(cast(tenant_code as TEXT), '') || '-' || coalesce(cast(lower(contact_unique_id) as TEXT), '') as TEXT)), 
+            null
+        )::varchar(32) as k_contact,
+        {{ gen_skey('k_student_xyear') }},
+        api_year as school_year,
+        unioned.*
+        {{ extract_extension(model_name=this.name, flatten=True) }}
```
[the remark this answers is not in the document]
Comment (sleblanc23) [reply models/staging/edfi_3/stage/stg_ef3__student_contact_associations.sql:33]: I don't think so because the two `v_ext` columns have been unioned together

File: models/staging/edfi_3/stage/stg_ef3__student_contact_associations.sql [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    union 
+    select * from base_stu_parent
+),
+keyed as (
+    select 
+        {{ gen_skey('k_student') }},
+        -- we can't use the gen_skey macro here because we're bringing in the deprecated parents endpoint data, which contains a parentReference that won't work
+        iff(
+            contact_unique_id is not null, 
+            md5(cast(coalesce(cast(tenant_code as TEXT), '') || '-' || coalesce(cast(lower(contact_unique_id) as TEXT), '') as TEXT)),
```
[the remark this answers is not in the document]
Comment (sleblanc23) [reply models/staging/edfi_3/stage/stg_ef3__student_contact_associations.sql]: yep, good call! updated to use surrogate_key. and it shouldn't ever be null, that's a mistake on my part.

**Comment by sleblanc23** [comment]
@rlittle08 I added a variable that has to be set to True in implementations that want to use Contacts (to avoid a big breaking change). Current name is `edu:contacts:enabled` but I'm not totally sure it's the right choice.

File: models/staging/edfi_3/stage/stg_ef3__student_contact_associations.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (rlittle08) [reply models/staging/edfi_3/stage/stg_ef3__student_contact_associations.sql:29]: but I think the macro will only extract the extensions configured under `this.name`, which is `stg_ef3__student_contact_associations`

[the remark this answers is not in the document]
Comment (sleblanc23) [reply models/staging/edfi_3/stage/stg_ef3__student_contact_associations.sql]: the macro can take a list of model names, so I added extensions configured under parents: `{{ extract_extension([model_name=this.name, 'stg_ef3__parents'], flatten=True) }}`

I don't have an implementation with extensions to test this on, but it should work based on the documentation of the macro

**Comment by rlittle08** [comment]
when we merge this, add to migration notes that implementations should copy in this codegen update: https://github.com/edanalytics/edu_project_template/pull/40/files
